### DISCUSSION TITLE
Activating imx415 for ubuntu build orangepi 5

### DISCUSSION THREAD

**Discussion by Abhra121** [Joshua-Riek/ubuntu-rockchip#477] [General]
Can you please say what are the changes to make to activate imx415 for the ubuntu build for ornagepi5 board for cam1 port

**Comment by BeckMeephanlom** [comment]
#This is very details for beginer who want to be expert. 
1. get your unbantu 22 here is stable version according your board https://joshua-riek.github.io/ubuntu-rockchip-download/	
2. You need to use RKDevtoll to Install it
3. update upgrade and install I2C bus and V4L2 
4. Reboot
5. Clone build sheel here  https://github.com/Joshua-Riek/ubuntu-rockchip	
6. Fork this into your account https://github.com/Joshua-Riek/linux-rockchip 
7. Update as below before start to build 
   	8.1 sudo apt update
	8.2 sudo apt install build-essential debhelper dh-autoreconf gawk
	8.3 sudo apt install gcc-aarch64-linux-gnu
	8.4 sudo apt install libssl-dev
	8.5 sudo apt install flex
        8.6 sudo apt install bison
	8.7 cd ubuntu-rockchip
	8.8 sudo nano config/suites/noble.sh
		export RELASE_NAME="Ubuntu 24.04 LTS (Noble Nombat)"
		export RELASE_VERSION="24.04"

		#export KERNEL_REPO="https://github.com/Joshua-Riek/linux-rockchip.git"
		export KERNEL_REPO="https://github.com/YOURNAME/linux-rockchip.git"			

		export KERNEL_BRANCH="noble"
		export KERNEL_FLAVOR="rockchip"

		export EXTRA_PPAS="jjriek/rockchip jjriek/rockchip-multimedia"
	8.9  Sudo reboot
		8.11 sudo ./build.sh --board=orangepi-cm5 --suite=noble --flavor=desktop --kernel-only			#This support to link up to github not compile on local disk 24.10	
		Required arguments:
  			-b, --board=BOARD      target board
  			-s, --suite=SUITE      ubuntu suite
  			-f, --flavor=FLAVOR    ubuntu flavor

		Optional arguments:
  			-h,  --help            show this help message and exit
  			-c,  --clean           clean the build directory
  			-ko, --kernel-only     only compile the kernel
  			-uo, --uboot-only      only compile uboot
  			-ro, --rootfs-only     only build rootfs
  			-l,  --launchpad       use kernel and uboot from launchpad repo
  			-v,  --verbose         increase the verbosity of the bash script


	8.12 sudo ./build.sh --board=orangepi-cm5 --suite=noble --flavor=desktop --kernel-only | tee build-log.txt     #incase of error can find and fix it
	
		#Finally when complete you could see this 
		dh_md5sums -plinux-headers-6.1.0-1027-rockchip
		dh_builddeb -plinux-headers-6.1.0-1027-rockchip
		dpkg-deb: building package 'linux-headers-6.1.0-1027-rockchip' in '../linux-headers-6.1.0-1027-rockchip_6.1.0-1027.27_arm64.deb'.
	
9. 	sudo dpkg -i ~/ubuntu-rockchip/build/linux-*.deb     #install above package to your system
           #at the end you could see this message
           		/etc/kernel/postinst.d/zz-u-boot-menu:
		P: Checking for EXTLINUX directory... found.
		P: Writing config for vmlinuz-6.1.0-1027-rockchip...
		P: Writing config for vmlinuz-6.1.0-1025-rockchip...
		P: Updating /boot/extlinux/extlinux.conf...
10. sudo nano /etc/default/u-boot	               You needto enable Uboot as below:

		## /etc/default/u-boot - configuration file for u-boot-update(8)

		U_BOOT_UPDATE="true"

		#U_BOOT_ALTERNATIVES="default recovery"
		U_BOOT_DEFAULT="l0"
		#U_BOOT_DEFAULT="l1"
		#U_BOOT_PROMPT="1"
		#U_BOOT_ENTRIES="all"
		#U_BOOT_MENU_LABEL="Debian GNU/Linux"
		#U_BOOT_PARAMETERS="ro earlycon"
		#U_BOOT_ROOT=""
		#U_BOOT_TIMEOUT="50"
		#U_BOOT_FDT=""
		U_BOOT_FDT_DIR="/lib/firmware/"
		U_BOOT_FDT_OVERLAYS="6.1.0-1027-rockchip/device-tree/rockchip/overlay/orangepi-cm5-cam1.dtbo"		//check it
		U_BOOT_FDT_OVERLAYS_DIR="/usr/lib/firmware/6.1.0-1027-rockchip/device-tree/rockchip/"
		#U_BOOT_SYNC_DTBS="false"
11.   sudo u-boot-update        #do this only when you update or change Uboot 
12.  sudo nano /boot/extlinux/extlinux.conf	                     #you will see 6.1.0-1027-rockchip become default
      
		## /boot/extlinux/extlinux.conf
		##
		## IMPORTANT WARNING
		##
		## The configuration of this file is generated automatically.
		## Do not edit this file manually, use: u-boot-update

		default l0											#default Kernel of system
		

		menu title U-Boot menu
		prompt 1
		timeout 20


		label l0
			menu label Ubuntu 24.04.2 LTS 6.1.0-1025-rockchip
			linux /boot/vmlinuz-6.1.0-1025-rockchip
			initrd /boot/initrd.img-6.1.0-1025-rockchip
			fdtdir /lib/firmware/6.1.0-1025-rockchip/device-tree/
	
			append root=UUID=eb11c793-66ca-4a6d-ad92-240381c70b69 rootwait rw console=ttyS2,1500000 console=tty1 cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory quiet splash plymouth.ignore-serial-consoles

		label l0r
			menu label Ubuntu 24.04.2 LTS 6.1.0-1025-rockchip (rescue target)
			linux /boot/vmlinuz-6.1.0-1025-rockchip
			initrd /boot/initrd.img-6.1.0-1025-rockchip
			fdtdir /lib/firmware/6.1.0-1025-rockchip/device-tree/
			append root=UUID=eb11c793-66ca-4a6d-ad92-240381c70b69 rootwait rw console=ttyS2,1500000 console=tty1 cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory splash plymouth.ignore-serial-consoles single
	
		#THIS LINE ADD UP AFTER DPKG
		label l1
        		menu label Ubuntu 24.04.2 LTS 6.1.0-1025-rockchip
        		linux /boot/vmlinuz-6.1.0-1025-rockchip
        		initrd /boot/initrd.img-6.1.0-1025-rockchip
        		fdtdir /lib/firmware/6.1.0-1025-rockchip/device-tree/

        		append root=UUID=eb11c793-66ca-4a6d-ad92-240381c70b69 rootwait rw console=ttyS2,1500000 console=tty1 cgroup_enable=cpuset cgroup_me>

		label l1r
        		menu label Ubuntu 24.04.2 LTS 6.1.0-1025-rockchip (rescue target)
        		linux /boot/vmlinuz-6.1.0-1025-rockchip
        		initrd /boot/initrd.img-6.1.0-1025-rockchip
        		fdtdir /lib/firmware/6.1.0-1025-rockchip/device-tree/
        		append root=UUID=eb11c793-66ca-4a6d-ad92-240381c70b69 rootwait rw console=ttyS2,1500000 console=tty1 cgroup_enable=cpuset cgroup_me>
 13.  sudo reboot
 14. hook up iMX415 15P 2 Lane to CAM1 Connector now use I2C Cammand to check. I got WaveShare IMX415 21Pin, but you can purchase FCC 15P as same as RPI 5M Camera Cable to convert it. This Orange Pi and Raspberry Pi are the same pinning.
 15. when it detect 0x37 adress So connection ready ...
 16. This is build in Kernel  and it will make you alway build the whole kernel not like module
      	CONFIG_VIDEO_IMX415=y							# y = build in Kernel
	CONFIG_VIDEO_IMX464=y
	# CONFIG_VIDEO_IMX492 is not set
	CONFIG_VIDEO_IMX498=y
	CONFIG_VIDEO_IMX577=m							# Module 
	CONFIG_VIDEO_IMX586=m
	CONFIG_VIDEO_IMX708=y
17. BeckMeephanlom/linux-rockchip/
      17.1 Copy blob/noble/drivers/media/i2c/imx415.c       which i explain everything and you can modify what ever you want but do not               use greater 2079Mbps it will be overheaded 2.5Gbps. 
       17. 2. Copy arch/arm64/boot/dts/rockchip/overlay/orangepi-cm5-cam1.dts
       17.3  Copy arch/arm64/boot/dts/rockchip/rk3588s-orangepi-cm5-camera1.dtsi
       17.4  Ensure you edit noble.sh to link to your forked 
18. start compile, once finish depackage and reboot
19.  dmesg | grep -i imx415  you could see this message
20. v4l2-ctl --list-devices
 		rkisp-statistics (platform: rkisp):
        		/dev/video18
        		/dev/video19

		rkcif (platform:rkcif-mipi-lvds2):
        		/dev/video0					
        		/dev/video1
        		/dev/video2
        		/dev/video3
        		/dev/video4
        		/dev/video5
        		/dev/video6
        		/dev/video7
        		/dev/video8
        		/dev/video9
        		/dev/video10
        		/dev/media0

		rkisp_mainpath (platform:rkisp0-vir0):--------------------------------Here video to display
        		/dev/video11						<<<<<<This can change anytime 	for me video22	
        		/dev/video12
        		/dev/video13
        		/dev/video14
        		/dev/video15
        		/dev/video16
        		/dev/video17
        		/dev/media1
#IMX415
21.  v4l2-ctl --device=/dev/video22 --all
Driver Info:
        Driver name      : rkisp_v6
        Card type        : rkisp_mainpath
        Bus info         : platform:rkisp0-vir0
        Driver version   : 2.6.0
        Capabilities     : 0x84201000
                Video Capture Multiplanar
                Streaming
                Extended Pix Format
                Device Capabilities
        Device Caps      : 0x04201000
                Video Capture Multiplanar
                Streaming
                Extended Pix Format
Media Driver Info:
        Driver name      : rkisp0-vir0
        Model            : rkisp0
        Serial           :
        Bus info         : platform:rkisp0-vir0
        Media version    : 6.1.75
        Hardware revision: 0x00000000 (0)
        Driver version   : 6.1.75
Interface Info:
        ID               : 0x03000007
        Type             : V4L Video
Entity Info:
        ID               : 0x00000006 (6)
        Name             : rkisp_mainpath
        Function         : V4L2 I/O
        Pad 0x01000009   : 0: Sink
          Link 0x0200000a: from remote pad 0x1000004 of entity 'rkisp-isp-subdev' (Unknown V4L2 Sub-Device): Data, Enabled
Priority: 2
Format Video Capture Multiplanar:
        Width/Height      : 3840/2160
        Pixel Format      : 'NV12' (Y/UV 4:2:0)
        Field             : None
        Number of planes  : 1
        Flags             :
        Colorspace        : Default
        Transfer Function : Default
        YCbCr/HSV Encoding: Default
        Quantization      : Full Range
        Plane 0           :
           Bytes per Line : 3840
           Size Image     : 12441600
Selection Video Capture: crop, Left 0, Top 0, Width 3840, Height 2160, Flags:
Selection Video Capture: crop_bounds, Left 0, Top 0, Width 3840, Height 2160, Flags:
Selection Video Output: crop, Left 0, Top 0, Width 3840, Height 2160, Flags:
Selection Video Output: crop_bounds, Left 0, Top 0, Width 3840, Height 2160, Flags:

Image Processing Controls

                     pixel_rate 0x009f0902 (int64)  : min=0 max=1000000000 step=1 default=1000000000 value=148666664 flags=read-only, volatile
  22.  gst-launch-1.0 v4l2src device=/dev/video22 ! video/x-raw,format=NV12,width=3840,height=2160 ! videoconvert ! autovideosink		#IMX415    
#you could see darky video on you Opi Monitor becuase need to setup init value for it
#########This is 3840x2160 Pixel, You need to config width and Height to meet your Desktop Resolution to avoid #STREAMING STOPPED!!!!###########
23.  v4l2-ctl -d /dev/video22 --stream-mmap=3 --stream-count=10000       #test Frame Rate count
24. v4l2-ctl -d /dev/v4l-subdev2 --list-ctrls      #show control lists 

        white_balance_automatic 0x0098090c (bool)   : default=0 value=0
                       exposure 0x00980911 (int)    : min=4 max=1964 step=1 default=1000 value=89
                 gain_automatic 0x00980912 (bool)   : default=0 value=0
                horizontal_flip 0x00980914 (bool)   : default=0 value=0 flags=modify-layout
                  vertical_flip 0x00980915 (bool)   : default=0 value=0 flags=modify-layout

Camera Controls

                  auto_exposure 0x009a0901 (menu)   : min=0 max=1 default=1 value=1 (Manual Mode)

Image Source Controls

![Orange Pi CM5](https://github.com/user-attachments/assets/62978fa6-d2a9-4f28-9aa0-6c8552486ff3)

**Comment by BeckMeephanlom** [comment]
This is you can turn it to 4 Lane for your prototype.
![Uploading Orange Pi CM5.png…]()

**Comment by Sillister** [comment]
Hey guys, I have followed your guide an created a script for the Orange Pi 5 Plus (RK3588) paired with the Tachyn Labs adater to use a Waveshare imx415 camera.
Feel free to use it, it should automatically install an configure everything:

#!/bin/bash
set -e

echo "==============================================================="
echo " Orange Pi 5 Plus - IMX415 Camera Setup Script"
echo "==============================================================="

# 1. Update system and install required tools (including GStreamer for video/photo)
echo "[1/8] Installing dependencies, Rockchip ISP engine, and GStreamer..."
sudo apt-get update -y
sudo apt-get install -y build-essential git device-tree-compiler v4l-utils \
    camera-engine-rkaiq i2c-tools \
    gstreamer1.0-tools gstreamer1.0-plugins-base gstreamer1.0-plugins-good \
    gstreamer1.0-plugins-bad gstreamer1.0-plugins-ugly

# 2. Setup Workspace & Shallow Clone
WORKSPACE="$HOME/opi5plus-cam-setup"
mkdir -p "$WORKSPACE"
cd "$WORKSPACE"

echo "[2/8] Fetching Linux headers (shallow clone for maximum speed)..."
if [ ! -d "linux-rockchip" ]; then
    git clone --depth=1 -b noble https://github.com/Joshua-Riek/linux-rockchip.git
fi
cd linux-rockchip

# 3. Generate Device Tree Files
echo "[3/8] Generating Device Tree Overlay for Orange Pi 5 Plus (MIPI CSI0 / I2C3)..."
mkdir -p arch/arm64/boot/dts/rockchip/overlay/

cat << 'EOF' > arch/arm64/boot/dts/rockchip/overlay/rk3588-orangepi-5-plus-imx415.dtsi
// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
&csi2_dphy0_hw { status = "okay"; };
&csi2_dphy0 {
    status = "okay";
    ports {
        #address-cells = <1>;
        #size-cells = <0>;
        port@0 {
            reg = <0>;
            #address-cells = <1>;
            #size-cells = <0>;
            mipidphy0_in_ucam0: endpoint@1 {
                reg = <1>;
                remote-endpoint = <&imx415_1_out1>;
                data-lanes = <1 2 3 4>;
            };
        };
        port@1 {
            reg = <1>;
            #address-cells = <1>;
            #size-cells = <0>;
            csidphy0_out: endpoint@0 {
                reg = <0>;
                remote-endpoint = <&mipi2_csi2_input>;
            };
        };
    };
};
&i2c3 {
    status = "okay";
    pinctrl-names = "default";
    pinctrl-0 = <&i2c3m0_xfer>;
    imx415_1: imx415-1@37 {
        status = "okay";
        compatible = "sony,imx415";
        reg = <0x37>;
        clocks = <&cru CLK_MIPI_CAMARAOUT_M3>;
        clock-names = "xvclk";
        pinctrl-names = "default";
        pinctrl-0 = <&mipim0_camera3_clk>;
        power-domains = <&power RK3588_PD_VI>;
        pwdn-gpios = <&gpio1 RK_PD0 GPIO_ACTIVE_LOW>;
        rockchip,camera-module-index = <0>;
        rockchip,camera-module-facing = "back";
        rockchip,camera-module-name = "WaveShare-IMX415";
        rockchip,camera-module-lens-name = "default";
        port {
            imx415_1_out1: endpoint {
                remote-endpoint = <&mipidphy0_in_ucam0>;
                data-lanes = <1 2 3 4>;
            };
        };
    };
};
&mipi2_csi2 {
    status = "okay";
    ports {
        #address-cells = <1>;
        #size-cells = <0>;
        port@0 {
            reg = <0>;
            #address-cells = <1>;
            #size-cells = <0>;
            mipi2_csi2_input: endpoint@1 {
                reg = <1>;
                remote-endpoint = <&csidphy0_out>;
            };
        };
        port@1 {
            reg = <1>;
            #address-cells = <1>;
            #size-cells = <0>;
            mipi2_csi2_output: endpoint@0 {
                reg = <0>;
                remote-endpoint = <&cif_mipi_in2>;
            };
        };
    };
};
&rkcif_mipi_lvds2 {
    status = "okay";
    port { cif_mipi_in2: endpoint { remote-endpoint = <&mipi2_csi2_output>; }; };
};
&rkcif_mipi_lvds2_sditf {
    status = "okay";
    port { mipi2_lvds_sditf: endpoint { remote-endpoint = <&isp0_vir0>; }; };
};
&rkisp0_vir0 {
    status = "okay";
    port {
        #address-cells = <1>;
        #size-cells = <0>;
        isp0_vir0: endpoint@0 { reg = <0>; remote-endpoint = <&mipi2_lvds_sditf>; };
    };
};
&rkcif { status = "okay"; };
&rkcif_mmu { status = "okay"; };
&isp0_mmu { status = "okay"; };
&rkisp0 { status = "okay"; };
EOF

cat << 'EOF' > arch/arm64/boot/dts/rockchip/overlay/orangepi-5-plus-imx415.dts
/dts-v1/;
/plugin/;
#include <dt-bindings/gpio/gpio.h>
#include <dt-bindings/clock/rk3588-cru.h>
#include <dt-bindings/pinctrl/rockchip.h>
#include <dt-bindings/power/rk3588-power.h>
#include "rk3588-orangepi-5-plus-imx415.dtsi"
/ { compatible = "rockchip,rk3588"; };
EOF

# 4. Compile Overlay
echo "[4/8] Compiling the Device Tree Overlay (.dtbo)..."
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- rockchip_linux_defconfig
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- rockchip/overlay/orangepi-5-plus-imx415.dtbo

# 5. Copy to Firmware Directory dynamically based on current kernel version
echo "[5/8] Installing Overlay into the system firmware..."
KERNEL_VER=$(uname -r)
TARGET_DIR="/lib/firmware/$KERNEL_VER/device-tree/rockchip/overlay"
sudo mkdir -p "$TARGET_DIR"
sudo cp arch/arm64/boot/dts/rockchip/overlay/orangepi-5-plus-imx415.dtbo "$TARGET_DIR/"

# 6. Patch extlinux.conf dynamically
echo "[6/8] Patching bootloader (extlinux.conf) safely..."
EXTLINUX="/boot/extlinux/extlinux.conf"
DTBO_STR="        fdtoverlays /lib/firmware/$KERNEL_VER/device-tree/rockchip/overlay/orangepi-5-plus-imx415.dtbo"

if grep -q "orangepi-5-plus-imx415.dtbo" "$EXTLINUX"; then
    echo "Overlay is already present in extlinux.conf. Skipping patch."
else
    awk -v dtbo="$DTBO_STR" '/fdtdir/ && !done { print; print dtbo; done=1; next } 1' "$EXTLINUX" > temp.conf
    sudo mv temp.conf "$EXTLINUX"
fi

# 7. Patch Auto-Exposure IQ profiles
echo "[7/8] Patching Rockchip AIQ profiles for auto-exposure..."
if [ -f "/etc/iqfiles/imx415_RADXA-CAMERA-4K_DEFAULT.json" ]; then
    sudo cp /etc/iqfiles/imx415_RADXA-CAMERA-4K_DEFAULT.json /etc/iqfiles/imx415_WaveShare-IMX415_default.json
    echo "IQ profile configured successfully."
else
    echo "WARNING: Base IQ file not found. Auto-exposure might need manual tuning."
fi

# 8. Start ISP Services
echo "[8/8] Enabling Rockchip ISP AIQ engine (rkaiq_3A)..."
sudo systemctl enable rkaiq_3A || true
sudo systemctl restart rkaiq_3A || true

echo "==============================================================="
echo " INSTALLATION COMPLETE! 🎉"
echo " Your IMX415 camera is fully configured with GStreamer support"
echo " and auto-exposure tuning."
echo " "
echo " Please restart your board now:"
echo " sudo reboot"
echo "==============================================================="

**Reply by parshukov** [reply]
Hi! I've done the same with my orangepi 5b and it was not succeed... I have imx415 camera from [aliexpress](https://aliexpress.ru/item/1005009085172042.html?businessType=ProductDetail&sku_id=12000047855492383&spreadType=socialShare&srcSns=sns_More&utm_medium=sharing)

Any advices please...
